src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs

```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
             writer.IncreaseIndent();
             foreach (var mappedType in codeElement.OriginalMethodParentClass.DiscriminatorInformation.DiscriminatorMappings)
             {
-                var typeName = conventions.GetTypeString(mappedType.Value, codeElement, false, writer);
                 writer.WriteLine($"case \"{mappedType.Key}\":");
                 writer.IncreaseIndent();
-                writer.WriteLine($"return new {typeName.ToFirstCharacterUpperCase()}();");
+                writer.WriteLine($"return {getDeserializationFunction(codeElement, mappedType.Value.Name.ToFirstCharacterUpperCase())};");
                 writer.DecreaseIndent();
             }
             writer.CloseBlock();
             writer.CloseBlock();
             writer.CloseBlock();
         }
+        var s = getDeserializationFunction(codeElement, returnType);
+        writer.WriteLine($"return {s.ToFirstCharacterLowerCase()};");
+    }
+
+    private string getDeserializationFunction(CodeElement codeElement, string returnType)
+    {
+        if (codeElement.Parent is not CodeNamespace codeNamespace)
+        {
+            throw new InvalidOperationException($"{codeElement.Name} does not have a parent namespace");
+        }
+        var parent = codeNamespace.FindChildByName<CodeFunction>($"deserializeInto{returnType}");
+
+        return conventions.GetTypeString(new CodeType { TypeDefinition = parent }, codeElement, false);
+    }
+
+    private void WriteSerializerFunction(CodeFunction codeElement, LanguageWriter writer)
+    {
+        var param = codeElement.OriginalLocalMethod.Parameters.FirstOrDefault(x => ((CodeType)x.Type).TypeDefinition is CodeInterface);
```

Comment:
this will throw if it encounters a type that's a composed type. switch to type assertion instead of cast